azure-iot-device/azure/iot/device/common/pipeline/operation_flow.py

```diff
@@ -0,0 +1,243 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import sys
+
+from six.moves import queue
+
+logger = logging.getLogger(__name__)
+
+
+def run_ops_in_serial(stage, *args, **kwargs):
+    """
+    Run the operations passed in *args in a serial manner, such that each operation waits for the
+    previous operation to be completed before running.
+
+    In normal operation, the first operation will be passed down to the next stage to run.  That
+    stage will process it, pass it down, or whatever needs to be done to handle that operation.
+    After that operation travels all the way down the pipeline and the result of the operation is
+    returned via the operation's callback, the next operation will be passed down and execute.
+    When the second operation's callback is called, the third operation will begin, and so on.
+    After all of the operations run, the callback passed into run_ops_serial will be called to
+    return the "final result" back to the caller.
+
+    The finally_op parameter is used to provide an operation that will always be run as the last
+    operation, even if one of the earlier operation fails.
+
+    If an operation fails, the operations in *args that follow will not be run. However, the
+    operation passed using the finally_op will be run.
+
+    After all operations and the finally_op (if provided) are run, then the function passed in
+    the callback parameter will be called, passing the last operation executed (in the event of
+    success) or the failing operation (in the event of failure).
+
+    :param PipelineStage stage: stage to run these operations on
+    :param PipelineOperation[] *args: Operations to run serially.  These operations will be run
+      one-after-another which each operation only running after the previous operations have all
+      run.  If one of these operations fail, all subsequent operations will not execute and the
+      failed operation will be passed as a parameter to the callback with the error attribute
+      set to the cause of the failure.
+    :param PipelineOperation finally_op: Operation to run last.  This will be called regardless
+      of success or failure of the operations passed in *args
+    :param Function callback: Callback to execute when all operations have completed.  The last
+      operation executed will be passed as the parameter to this function.
+    """
+    worklist = queue.Queue()
+    for arg in args:
+        worklist.put_nowait(arg)
+    finally_op = kwargs.get("finally_op", None)
+    callback = kwargs.get("callback", None)
+
+    if len([x for x in kwargs.keys() if x not in ["finally_op", "callback"]]):
+        raise TypeError("finally_op and callback are the only allowed keyword args")
```

Comment:
>TypeError [](start = 14, length = 9)

Is this kind of kwarg-checking standard practice? I'm not honestly sure.

Reply:
If you specify kwargs in your method prototype, this check happens automatically.  In this function, we can't specify it in the prototype, so we have to check manually.  Ideally, we would define this function as def run_ops_in_serial(stage, *args, callback=None, finally_op=None), and this works in newer python versions but not in older ones.  If we did prototype this way, we would get an error if the user tried to use an invalid keyword.  I'm just replicating this behavior.

---
In reply to: [293450525](https://github.com/Azure/azure-iot-sdk-python-preview/pull/106#discussion_r293450525) [](ancestors = 293450525)